// @tsed/cli-plugin-typeorm/lib/commands/TypeORMCmd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeORMCmd = void 0;
const tslib_1 = require("tslib");
const cli_core_1 = require("@tsed/cli-core");
const CliTypeORM_1 = require("../services/CliTypeORM");
let TypeORMCmd = class TypeORMCmd {
    async $exec(ctx) {
        return [
            {
                title: `Run TypeORM CLI ${ctx.command}`,
                task: () => this.cliTypeORM.run(ctx.command, ctx.rawArgs)
            }
        ];
    }
};
tslib_1.__decorate([
    cli_core_1.Inject(),
    tslib_1.__metadata("design:type", CliTypeORM_1.CliTypeORM)
], TypeORMCmd.prototype, "cliTypeORM", void 0);
TypeORMCmd = tslib_1.__decorate([
    cli_core_1.Command({
        name: "typeorm",
        description: "Run a typeorm command",
        args: {
            command: {
                description: "The typeorm command",
                type: String,
                required: true
            }
        },
        options: {},
        allowUnknownOption: true
    })
], TypeORMCmd);
exports.TypeORMCmd = TypeORMCmd;
