// @tsed/common/lib/platform/PlatformModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformModule = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const mvc_1 = require("../mvc");
const PlatformControllerBuilder_1 = require("./builder/PlatformControllerBuilder");
const Platform_1 = require("./services/Platform");
const PlatformRouter_1 = require("./services/PlatformRouter");
/**
 * @ignore
 */
let PlatformModule = class PlatformModule {
    constructor(injector, platform) {
        this.injector = injector;
        this.platform = platform;
        this.createRoutersFromControllers();
    }
    /**
     * Create routers from the collected controllers
     */
    createRoutersFromControllers() {
        const { injector } = this;
        return injector
            .getProviders(di_1.ProviderType.CONTROLLER)
            .map((provider) => {
            provider.setRouter(PlatformRouter_1.PlatformRouter.create(injector, provider.routerOptions));
            if (!provider.hasParent()) {
                return new PlatformControllerBuilder_1.PlatformControllerBuilder(provider).build(injector);
            }
        })
            .filter(Boolean);
    }
    $$loadRoutes() {
        const routes = this.injector.settings.get("routes");
        this.platform.addRoutes(routes);
    }
};
PlatformModule = tslib_1.__decorate([
    di_1.Module({
        imports: [di_1.InjectorService, mvc_1.ConverterService, Platform_1.Platform]
    }),
    tslib_1.__metadata("design:paramtypes", [di_1.InjectorService, Platform_1.Platform])
], PlatformModule);
exports.PlatformModule = PlatformModule;
